Extra functionality in net.lua, full battery notification, a new widget for monitoring external drives

cool-cool-sweat opened the pull request:
The new helper function helpers.line_callback gets used in net.lua so that all network interfaces except loopback ones get added to net.iface, even if they don't have an active connection. This makes the widget more flexible since it can now detect changes in the interface after the widget's been loaded. This is useful if for example the widget's loaded while the ethernet's unconnected and then later the user switches to ethernet.

helpers.async and helpers.async_with_shell can now return the process' exit code, so widgets can use it in if statements. extdrive.lua makes use of that

lcpz commented on widget/net.lua:
```diff
@@ -9,28 +9,30 @@
+    local units     = args.units or 1024 -- KB
+    local notify    = args.notify or "on"
+    local wifi_info = args.wifi_info or "off"
+    local eth_info  = args.eth_info or "off"
```
The variables `{wifi,eth}_info` needs to be documented in the wiki. The notes you wrote in this commit description should also be reported.

lcpz commented on widget/net.lua:
```diff
@@ -66,9 +68,56 @@ local function factory(args)
             dev_now.last_t   = now_t
             dev_now.last_r   = now_r
 
+	    if wifi_info == "on" and os.execute(string.format("ls /sys/class/net/%s/wireless", dev)) and string.match(dev_now.carrier, "1") then
```
Indent here.

lcpz commented on widget/net.lua:
```diff
@@ -66,9 +68,56 @@ local function factory(args)
+	    if wifi_info == "on" and os.execute(string.format("ls /sys/class/net/%s/wireless", dev)) and string.match(dev_now.carrier, "1") then
+            dev_now.wifi = true
+
+            helpers.line_callback(string.format("iw dev %s link", dev), function(line)
```
This process is being respawned at each `net.update()` timeout. Since it is asynchronous, it will likely cause a memory leak.

lcpz commented on widget/net.lua:
```diff
@@ -66,9 +68,56 @@ local function factory(args)
+            dev_now.bitrate = bitrate
+        end
+
+	    if eth_info == "on" and not os.execute(string.format("ls /sys/class/net/%s/wireless", dev)) and string.match(dev_now.carrier, "1") then
```
Indent here.

lcpz commented on widget/net.lua:
```diff
@@ -66,9 +68,56 @@ local function factory(args)
+	    end
+
+	    if (eth_info == "on" or wifi_info == "on") and string.match(dev_now.carrier, "1") then
+            helpers.line_callback("ip neigh", function(line)
```
See line 74.

lcpz commented on widget/net.lua:
```diff
@@ -66,9 +68,56 @@ local function factory(args)
+                if net_mac_parse then net_mac = net_mac_parse end
+            end)
+
+            helpers.line_callback("ip addr show " .. dev, function(line)
```
See line 74.

lcpz commented on helpers.lua:
```diff
@@ -5,7 +5,7 @@
 --]]
 
-local spawn = require("awful.spawn")
+local spawn      = require("awful.spawn")
```
This is a good commit. If you separate it from this PR, I can merge it.

cool-cool-sweat replied:
Just the change to line 8 or you mean all the changes to the file?

lcpz replied:
All changes to the file.

lcpz commented on widget/bat.lua:
```diff
@@ -32,19 +32,19 @@ local function factory(args)
-        timeout = 15,
-        fg      = "#000000",
-        bg      = "#FFFFFF"
+        preset = naughty.config.presets.critical,
```
For double safety, please check if this won't modify `naughty.config.presets.critical`.

cool-cool-sweat replied:
It doesn't, but this line doesn't seem to actually affect the notification's appearance at least for me. If you want I'll change it back to the old style or I could manually add  naughty.config.preset.critical's settings

lcpz replied:
> this line doesn't seem to actually affect the notification's appearance at least for me

Are you referring to the added line? If so, what's the purpose of it?

cool-cool-sweat replied:
The purpose was to apply the settings in naughty.config.presets.critical [https://awesomewm.org/doc/api/libraries/naughty.html#config.presets] to the notification, but that doesn't seem to actually work inside the widget

lcpz commented:
There seems to be a major flaw in your logic. The flexibility you want to introduce in the net widget would require a respawn of the same processes at each update timeout. This is likely to led to a memory saturation. You can probably verify it by just monitoring `top` output.

Overall, I want these widgets to be light and simple, but there are some elements I can directly merge, and others we can work on to get as close to your desideratum as possible. Please, read the review.

cool-cool-sweat commented:
I can't find a memory leak in top, and system memory use stays basically the same with the widget removed and with it working with `wifi_info` set to on. And I've used the widget like this all day without it gobbling memory. But if you prefer I'll undo the changes to net.lua

cool-cool-sweat commented:
Don't worry about it, I'm sure you have other things you'd like to do. I'll work on this some more in a bit. Thanks for your patience

lcpz commented:
Taking this repository to release 1.0 after 4+ years of git is surely a thing I want to do, so I'll definitely come back at this.

Thank you for your interest and work, it is appreciated. 😹

cool-cool-sweat commented:
I reverted the changes to the notification presets in bat.lua since that wasn't really working out and I changed net.lua. Now it just sees if an ethernet connection's active and doesn't gather any info about it, and the only info it gets about the wifi's the signal strength. I have a draft edit for net.lua's wiki page

lcpz commented:
I've merged all except 7e9700f. I believe you can just use a [watcher](https://awesomewm.org/doc/api/classes/awful.widget.watch.html) for that. Example snippet:
```lua
local extdrive = awful.widget.watch(
    { awful.util.shell, "-c", "findmnt -n -o TARGET /dev/sdb1" }, 5,
    function(widget, stdout)
        -- do something to markup widget
    end
)
```
Once you've done, you can PR your recipe to [awesome-www](https://github.com/awesomeWM/awesome-www).

Don't forget to update the wiki. Thank you for your work. 👍
